django/db/models/sql/datastructures.py

```diff
@@ -79,17 +82,31 @@ def as_sql(self, compiler, connection):
         qn = compiler.quote_name_unless_alias
         qn2 = connection.ops.quote_name
 
+        if hasattr(self.join_field, "get_joining_fields"):
+            join_fields = self.join_field.get_joining_fields()
+        else:
+            warnings.warn(
+                "The usage of get_joining_columns in Join is deprecated. Implement "
+                "get_joining_fields instead.",
+                RemovedInDjango60Warning,
+            )
```

Comment:
I think we should also deprecate `Join.get_reverse_joining_columns()` method, `Join.join_cols` attribute, and related methods in `ForeignObject`  and `ForeignObjectRel` and mark them for removal in Django 6.0. There is no need to maintain 2 APIs for the same, and we don't use them anywhere else (\cc @charettes).

Reply:
Also, we should add tests for all deprecation warnings.

Reply:
The `Join.join_cols` is used in `django.db.models.sql.compiler.SQLCompiler.find_ordering_name`, so we won't deprecate it I guess.

----

Do you have a tip on how to cleanly test the hereby warning? I'm not sure how to end up with a `join_field` that is not a `ForeignObject(Rel)` 🤔